lib/src/Base/Stat/LatentVariableModel.cxx

```diff
@@ -0,0 +1,441 @@
+  activeParameter.fill();
+  setActiveParameter(activeParameter);
+  updateLatentCovarianceMatrix();
+}
+
+/** Parameters constructor */
+LatentVariableModel::LatentVariableModel(const UnsignedInteger latent_dim,
+						 const UnsignedInteger n_levels)
+  : CovarianceModelImplementation(Point(1, 1.0), Point(1, 1.0))
+  , latent_dim_(latent_dim)
+  , n_levels_(n_levels)
+  , latCovMat_(n_levels_)
+  , latCovMod_(latent_dim_)
+{
+  if (latent_dim_ < 1) throw InvalidArgumentException(HERE) << "Error: the dimension of the latent space must be >= 1";
+  if (n_levels_ < 2) throw InvalidArgumentException(HERE) << "Error: the number of discrete levels must be >= 2";
+  activeLatentCoordinateDim_ = 1 + latent_dim_ * (n_levels_ - 2);
+  activeLatentVariables_ = Point(activeLatentCoordinateDim_,0.);
+  fullLatentVariables_ = Sample(n_levels_,latent_dim_);
+  Indices activeParameter = Indices(inputDimension_ + outputDimension_ + activeLatentCoordinateDim_);
+  activeParameter.fill();
+  setActiveParameter(activeParameter);
+  updateLatentCovarianceMatrix();
+}
+
+/* Virtual constructor */
+LatentVariableModel * LatentVariableModel::clone() const
+{
+  return new LatentVariableModel(*this);
+}
+
+/* Computation of the covariance  function */
+Scalar LatentVariableModel::computeAsScalar(const Scalar z1, const Scalar z2) const
+{
+  /*check z being one of the levels*/
+  bool isLevelz1 = false;
+  bool isLevelz2 = false;
+
+  for (UnsignedInteger i = 0; i < n_levels_; ++i)
+  {
+    if (z1 == i) 
+    {
+      isLevelz1 = true;
+      break;
+    }
+  }
+
+  for (UnsignedInteger i = 0; i < n_levels_; ++i)
+  {
+    if (z2 == i)
+    {
+      isLevelz2 = true;
+      break;
+    }
+  }
+
+  if (!isLevelz1 || !isLevelz2) throw InvalidArgumentException(HERE) << "Error: the input discrete variables values are not amongst the known levels";
+
+  Scalar cov = latCovMat_(z1,z2);
+
+  return cov;
+}
+
+Scalar LatentVariableModel::computeAsScalar(const Point & z1, const Point & z2) const
+{
+  return computeAsScalar(z1[0],z2[0]);
+}
+
+
+Scalar LatentVariableModel::computeAsScalar(const Collection<Scalar>::const_iterator & z1_begin,
+    const Collection<Scalar>::const_iterator & z2_begin) const
+{
+
+  Collection<Scalar>::const_iterator z1_it = z1_begin;
+  Collection<Scalar>::const_iterator z2_it = z2_begin;
+
+  /*check z being one of the levels*/
+  bool isLevelz1 = false;
+  bool isLevelz2 = false;
+
+  for (UnsignedInteger i = 0; i < n_levels_; ++i)
+  {
+    if (*z1_it == i)
+    {
+      isLevelz1 = true;
+      break;
+    }
+  }
+
+  for (UnsignedInteger i = 0; i < n_levels_; ++i)
+  {
+    if (*z2_it == i)
+    {
+      isLevelz2 = true;
+      break;
+    }
+  }
+
+  if (!isLevelz1 || !isLevelz2) throw InvalidArgumentException(HERE) << "Error: the input discrete variables values are not amongst the known levels";
+
+  Scalar cov = latCovMat_(*z1_it,*z2_it);
+  return cov;
+}
```

Comment:
if the model is 1-d only you can maybe avoid to duplicate code and call the previous version